### PR TITLE
contrib: update non-client integrations to use configured service name

### PR DESCRIPTION
**Pull request by knusbaum**
As part of our continued work to unify the concept of service and bring the various tracers
in line with each other, we want to have integrations use the tracer's service name, if one has been configured through the new `WithService` `StartOption` or through the `DD_SERVICE` environment variable. Otherwise, the integrations should continue to behave as they did before. To avoid conflicts, this currently will only apply to "non-client" integrations, that is integrations for packages that are not primarily clients for external services, similar to what was done in #603 

Integrations set their own service name regardless of what is configured in
the tracer. This commit updates non-client integrations to pick up the
tracer's service name if one has been configured with either the
`DD_SERVICE` environment variable, or the `WithService` `StartOption`. They
fall back to the previous default service name for the integration if a
service name has not been set, or if it has been set with
`WithServiceName`, for backwards-compatibility. `WithServiceName` has been
deprecated, but retains its original behavior.

Integrations can still be individually configured with their own service
names which will override the one configured in the tracer.

In order to do this, the tracer exposes its service name through the
internal/globalconfig package, when set with `DD_SERVICE` or `WithService`

### CODE REVIEW

File: contrib/Shopify/sarama/option.go [hunk 1]
```diff
@@ -7,15 +7,24 @@ package sarama
...
-	analyticsRate float64
+	consumerServiceName string
+	producerServiceName string
+	analyticsRate       float64
 }
 
 func defaults(cfg *config) {
-	cfg.serviceName = "kafka"
+	cfg.consumerServiceName = "kafka"
+	cfg.producerServiceName = globalconfig.ServiceName()
```
Comment (jdgumz) [inline contrib/Shopify/sarama/option.go]: I see, so we can't just have `serviceName` because this global config is used for wrapping both the consumer and producer clients?

I would have thought we'd have two separate `config` objects, one for the consumer and one for the producer.

Comment (jdgumz) [reply contrib/Shopify/sarama/option.go]: Also shouldn't the consumer use the globally configured service name? Since the consumer is likely the actual service running, whereas the producer is just a client?

File: contrib/confluentinc/confluent-kafka-go/kafka/option.go [hunk 2]
```diff
@@ -8,24 +8,33 @@ package kafka
...
 func newConfig(opts ...Option) *config {
 	cfg := &config{
-		serviceName: "kafka",
-		ctx:         context.Background(),
+		ctx: context.Background(),
 		// analyticsRate: globalconfig.AnalyticsRate(),
 		analyticsRate: math.NaN(),
 	}
+	cfg.consumerServiceName = "kafka"
+	cfg.producerServiceName = globalconfig.ServiceName()
```
Comment (jdgumz) [inline contrib/confluentinc/confluent-kafka-go/kafka/option.go]: As I said above I think these might need to be reversed?

File: contrib/google.golang.org/grpc.v12/grpc.go [hunk 3]
```diff
@@ -31,7 +32,11 @@ func UnaryServerInterceptor(opts ...InterceptorOption) grpc.UnaryServerIntercept
 		fn(cfg)
 	}
 	if cfg.serviceName == "" {
-		cfg.serviceName = "grpc.server"
+		cfg.serviceName = globalconfig.ServiceName()
+		if cfg.serviceName == "" ||
+			cfg.serviceName == tracer.DefaultServiceName {
+			cfg.serviceName = "grpc.server"
+		}
```
Comment (jdgumz) [inline contrib/google.golang.org/grpc.v12/grpc.go]: :+1: this looks right to me.

File: contrib/julienschmidt/httprouter/option.go [hunk 4]
```diff
@@ -23,7 +24,11 @@ type RouterOption func(*routerConfig)
 
 func defaults(cfg *routerConfig) {
 	cfg.analyticsRate = globalconfig.AnalyticsRate()
-	cfg.serviceName = "http.router"
+	cfg.serviceName = globalconfig.ServiceName()
+	if cfg.serviceName == "" ||
+		cfg.serviceName == tracer.DefaultServiceName {
+		cfg.serviceName = "http.router"
```
Comment (jdgumz) [inline contrib/julienschmidt/httprouter/option.go]: Interesting - this is only a server-side framework right? It's not something also used for making client HTTP calls to external services?

File: internal/globalconfig/globalconfig.go [hunk 5]
```diff
@@ -36,3 +37,17 @@ func SetAnalyticsRate(rate float64) {
...
+
+// ServiceName returns the global service name set for this application.
+func ServiceName() string {
+	cfg.mu.RLock()
+	defer cfg.mu.RUnlock()
+	return cfg.serviceName
+}
+
+// SetServiceName sets the global service name set for this application.
+func SetServiceName(name string) {
```
Comment (jdgumz) [inline internal/globalconfig/globalconfig.go]: Yep this API makes sense to me.

File: contrib/Shopify/sarama/option.go (hunk 1 above)
Comment (knusbaum) [reply contrib/Shopify/sarama/option.go]: Yes, they use the same config, so we have to differentiate.

The decision to do the producer and not the consumer was made by @brettlangdon. I can see it either way, and I could also imagine considering both clients to an external service. I'm happy with whatever we decide, but for the metrics tagging, we tagged the producer.

File: contrib/julienschmidt/httprouter/option.go (hunk 4 above)
Comment (knusbaum) [reply contrib/julienschmidt/httprouter/option.go]: That's right, this is only server-side.

File: internal/globalconfig/globalconfig.go (hunk 5 above)
Comment (gbbr) [reply internal/globalconfig/globalconfig.go]: Yeah this isn't great but it's the best we can do. Because it's `internal` it's slightly ok :)

File: ddtrace/tracer/option.go [hunk 6]
```diff
@@ -22,6 +22,11 @@ import (
 	"gopkg.in/DataDog/dd-trace-go.v1/internal/version"
 )
 
+// DefaultServiceName is the service name that will be given to the tracer
+// if one is not configured with WithServiceName or the DD_SERVICE
+// environment variable.
+var DefaultServiceName string = filepath.Base(os.Args[0])
```
Comment (gbbr) [inline ddtrace/tracer/option.go]: This is not ok. Global level exported variables can be dangerous, are prone to race conditions and users or foreign (even potentially malicious packages) can alter them. Read more about this theory [here](https://dave.cheney.net/2017/06/11/go-without-package-scoped-variables).

Why do you want to add this?

File: ddtrace/tracer/option.go [hunk 7]
```diff
@@ -111,7 +116,7 @@ func defaults(c *config) {
 	if v := os.Getenv("DD_SERVICE"); v != "" {
 		c.serviceName = v
 	} else {
-		c.serviceName = filepath.Base(os.Args[0])
+		c.serviceName = DefaultServiceName
```
Comment (gbbr) [inline ddtrace/tracer/option.go]: We should think of a good way to make sure that the fact that the service name exists in two places doesn't become confusing. One idea is to clearly document that the service in the `globalconfig` package refers strictly to integrations.

File: ddtrace/tracer/tracer.go [hunk 8]
```diff
@@ -172,7 +173,8 @@ func newUnstartedTracer(opts ...StartOption) *tracer {
 func newTracer(opts ...StartOption) *tracer {
 	t := newUnstartedTracer(opts...)
 	c := t.config
-	t.config.statsd.Incr("datadog.tracer.started", nil, 1)
+	globalconfig.SetServiceName(c.serviceName)
```
Comment (gbbr) [inline ddtrace/tracer/tracer.go]: Why is this not in the same place where all service setting related logic is?

File: ddtrace/tracer/option.go (hunk 6 above)
Comment (knusbaum) [reply ddtrace/tracer/option.go]: This *was* the way I was exposing the default, so integrations could determine whether the serviceName was the default service name or not. I could have made it a function instead to avoid the global exported var. It is no longer needed since we've slightly changed the design.

File: ddtrace/tracer/option.go (hunk 7 above)
Comment (knusbaum) [reply ddtrace/tracer/option.go]: This is also no longer an issue with the new implementation.

File: contrib/Shopify/sarama/option.go [hunk 9]
```diff
@@ -26,7 +33,8 @@ type Option func(cfg *config)
 // WithServiceName sets the given service name for the intercepted client.
 func WithServiceName(name string) Option {
 	return func(cfg *config) {
-		cfg.serviceName = name
+		cfg.consumerServiceName = name
+		cfg.producerServiceName = name
```
Comment (jdgumz) [inline contrib/Shopify/sarama/option.go]: Okay got it - this is the new change we discussed.

I'm wondering if we want both consumer and producer to be under the same service. It might make sense for queue-based applications to consider consumption/production as operations under that one service.

File: ddtrace/tracer/option.go [hunk 10]
```diff
@@ -186,16 +187,22 @@ func WithPropagator(p Propagator) StartOption {
...
 }
 
-// WithService sets the default service name to be used with the tracer.
+// WithService sets the default service name to be used with the tracer. It will also be
+// used as the service name for server and framework integrations
+// (e.g. http servers, grpc, etc.)
 func WithService(name string) StartOption {
 	return func(c *config) {
 		c.serviceName = name
+		globalconfig.SetServiceName(c.serviceName)
```
Comment (jdgumz) [inline ddtrace/tracer/option.go]: :+1: Okay makes sense. Basically the new functionality for "core service" comes from checking/using `globalconfig.ServiceName()`.

File: ddtrace/tracer/option_test.go [hunk 11]
```diff
@@ -158,3 +146,34 @@ func TestTracerOptionsDefaults(t *testing.T) {
...
+
+	t.Run("env-service", func(t *testing.T) {
+		os.Setenv("DD_SERVICE", "api-intake")
+		defer os.Unsetenv("DD_SERVICE")
+		assert := assert.New(t)
+		tracer := newTracer()
+
+		assert.Equal("api-intake", tracer.config.serviceName)
+		assert.Equal("api-intake", globalconfig.ServiceName())
+	})
```
Comment (jdgumz) [inline ddtrace/tracer/option_test.go]: Test case for both `DD_SERVICE` and using `WithServiceName`?

File: contrib/Shopify/sarama/option.go (hunk 9 above)
Comment (knusbaum) [reply contrib/Shopify/sarama/option.go]: @brettlangdon What do you think?

File: ddtrace/tracer/option_test.go [hunk 12]
```diff
@@ -158,3 +146,47 @@ func TestTracerOptionsDefaults(t *testing.T) {
...
+		os.Setenv("DD_SERVICE", "api-intake")
+		defer os.Unsetenv("DD_SERVICE")
+		assert := assert.New(t)
+
+		tracer := newTracer(
+			WithServiceName("api-intake"),
+		)
+
+		assert.Equal("api-intake", tracer.config.serviceName)
+		assert.Equal("", globalconfig.ServiceName())
```
Comment (jdgumz) [inline ddtrace/tracer/option_test.go]: Looks good!

File: contrib/Shopify/sarama/option.go [hunk 13]
```diff
@@ -7,15 +7,22 @@ package sarama
...
+	consumerServiceName string
+	producerServiceName string
+	analyticsRate       float64
 }
 
 func defaults(cfg *config) {
-	cfg.serviceName = "kafka"
+	cfg.consumerServiceName = "kafka"
+	cfg.producerServiceName = globalconfig.ServiceName()
+	if cfg.producerServiceName == "" {
```
Comment (gbbr) [inline contrib/Shopify/sarama/option.go]: Seems a bit backwards to me. Can we try this instead:
```go
cfg.consumerServiceName = "kafka"
cfg.producerServiceName = "kafka"
if svc := globalconfig.ServiceName(); svc != "" {
    cfg.producerServiceName = svc
}
```
What do you think? Better? Worse? Meah?

File: contrib/confluentinc/confluent-kafka-go/kafka/option.go [hunk 14]
```diff
@@ -8,24 +8,31 @@ package kafka
...
 
 func newConfig(opts ...Option) *config {
 	cfg := &config{
-		serviceName: "kafka",
-		ctx:         context.Background(),
+		ctx: context.Background(),
 		// analyticsRate: globalconfig.AnalyticsRate(),
 		analyticsRate: math.NaN(),
 	}
+	cfg.consumerServiceName = "kafka"
```
Comment (gbbr) [inline contrib/confluentinc/confluent-kafka-go/kafka/option.go]: Can we just declare the defaults above in the inline `cfg` structure declaration for both service names and then use the same as suggested above here?

```go
if svc := globalconfig.ServiceName(); svc != "" {
    cfg.producerServiceName = svc
}
```

File: contrib/go-chi/chi/option.go [hunk 15]
```diff
@@ -22,7 +22,10 @@ type config struct {
 type Option func(*config)
 
 func defaults(cfg *config) {
-	cfg.serviceName = "chi.router"
+	cfg.serviceName = globalconfig.ServiceName()
```
Comment (gbbr) [inline contrib/go-chi/chi/option.go]: Same here I guess. It's odd to me to assign something to a variable and then check if that variable is empty, when in fact we are actually checking the value of the assignment for empty. Does that even make sense?

File: contrib/google.golang.org/grpc.v12/grpc.go [hunk 16]
```diff
@@ -31,7 +32,10 @@ func UnaryServerInterceptor(opts ...InterceptorOption) grpc.UnaryServerIntercept
 		fn(cfg)
 	}
 	if cfg.serviceName == "" {
-		cfg.serviceName = "grpc.server"
+		cfg.serviceName = globalconfig.ServiceName()
```
Comment (gbbr) [inline contrib/google.golang.org/grpc.v12/grpc.go]: Same.

File: contrib/google.golang.org/grpc/option.go [hunk 17]
```diff
@@ -27,7 +29,10 @@ type config struct {
 
 func (cfg *config) serverServiceName() string {
 	if cfg.serviceName == "" {
-		return "grpc.server"
+		cfg.serviceName = globalconfig.ServiceName()
```
Comment (gbbr) [inline contrib/google.golang.org/grpc/option.go]: Here and all :)

File: ddtrace/tracer/option.go [hunk 18]
```diff
@@ -186,16 +187,23 @@ func WithPropagator(p Propagator) StartOption {
...
-// WithServiceName sets the default service name to be used with the tracer. It is
-// deprecated in favour of WithService and will be removed in the next major version.
+// WithServiceName sets the default service name to be used with the tracer. This will
+// not affect the service name used by any integrations. It is deprecated in favour of
+// WithService and will be removed in the next major version.
 func WithServiceName(name string) StartOption {
-	return WithService(name)
+	return func(c *config) {
+		c.serviceName = name
+		globalconfig.SetServiceName("")
```
Comment (gbbr) [inline ddtrace/tracer/option.go]: This seems hacky. Why do we need it? Let's drop it. We don't care if the user uses both of these options. The documentation should clearly explain the behaviour.

File: ddtrace/tracer/option.go [hunk 19]
```diff
@@ -186,16 +187,23 @@ func WithPropagator(p Propagator) StartOption {
 	}
 }
 
-// WithServiceName sets the default service name to be used with the tracer. It is
-// deprecated in favour of WithService and will be removed in the next major version.
+// WithServiceName sets the default service name to be used with the tracer. This will
+// not affect the service name used by any integrations. It is deprecated in favour of
+// WithService and will be removed in the next major version.
```
Comment (gbbr) [inline ddtrace/tracer/option.go]: I would like to suggest:
```go
// WithServiceName is deprecated. Please use WithService. 
// If you are using an older version and you are upgrading from WithServiceName
// to WithService, please note that WithService will determine server and framework
// integrations to share the service name. This provides a better experience of the
// product.
```
We should make sure to explain the same thing in the release notes, along with clear explanations about why the new form is better (@brettlangdon get ready to help with this 🙂)

File: ddtrace/tracer/option.go [hunk 20]
```diff
@@ -186,16 +187,23 @@ func WithPropagator(p Propagator) StartOption {
...
+	return func(c *config) {
+		c.serviceName = name
+		globalconfig.SetServiceName("")
+	}
 }
 
-// WithService sets the default service name to be used with the tracer.
+// WithService sets the default service name to be used with the tracer. It will also be
+// used as the service name for server and framework integrations
+// (e.g. http servers, grpc, etc.)
```
Comment (gbbr) [inline ddtrace/tracer/option.go]: Since we consider this behaviour normal and as it should be - would it be better to just:

```suggestion
// WithService sets the default service name for the program.
```

File: internal/globalconfig/globalconfig.go [hunk 21]
```diff
@@ -36,3 +37,17 @@ func SetAnalyticsRate(rate float64) {
 	cfg.analyticsRate = rate
 	cfg.mu.Unlock()
 }
+
+// ServiceName returns the global service name set for this application.
```
Comment (gbbr) [inline internal/globalconfig/globalconfig.go]:
```suggestion
// ServiceName returns the default service name used by non-client integrations such as servers and frameworks.
```
Right?

File: ddtrace/tracer/option.go (hunk 20 above)
Comment (knusbaum) [reply ddtrace/tracer/option.go]: Agreed. I went back and forth on this, but I'm ok being more terse.

File: contrib/go-chi/chi/option.go (hunk 15 above)
Comment (knusbaum) [reply contrib/go-chi/chi/option.go]: I can see it either way. If you *expect* the integration to get its name from `globalconfig.ServiceName()`, and consider `"chi.router"` a fallback, then this way makes sense. If you consider `"chi.router"` a default, which gets overridden by `globalconfig.ServiceName()`, then it makes sense to do it the way you suggest. Both have the same effect, it's just about how the reader interprets the intent. "Service name is the global service name. If it's empty, then use X" vs "Service name is X. If the user set a global name, then use that."

I will change it as you suggest, since I am not partial to one way or the other.

File: contrib/google.golang.org/grpc.v12/grpc.go (hunk 16 above)
Comment (knusbaum) [reply contrib/google.golang.org/grpc.v12/grpc.go]: 👍

File: contrib/google.golang.org/grpc/option.go (hunk 17 above)
Comment (knusbaum) [reply contrib/google.golang.org/grpc/option.go]: 👍

File: contrib/Shopify/sarama/option.go (hunk 13 above)
Comment (knusbaum) [reply contrib/Shopify/sarama/option.go]: 👍 Responded below.

File: ddtrace/tracer/option.go (hunk 19 above)
Comment (knusbaum) [reply ddtrace/tracer/option.go]: Made a minor modification, but 👍

File: ddtrace/tracer/option.go (hunk 18 above)
Comment (jdgumz) [reply ddtrace/tracer/option.go]: I think we said we would print a warning instead? Although I'm not sure if that will be visible enough to register.

Comment (gbbr) [reply ddtrace/tracer/option.go]: I'm on the fence. I still think we should remove it. `DD_SERVICE` is equivalent to using the `WithService` option. The counterpart for `WithServiceName` would be `DD_SERVICE_NAME` (which we don't have). I doubt someone will have `DD_SERVICE` different from `DD_SERVICE_NAME`.

File: ddtrace/tracer/option.go [hunk 22]
```diff
@@ -110,6 +110,7 @@ func defaults(c *config) {
 	}
 	if v := os.Getenv("DD_SERVICE"); v != "" {
 		c.serviceName = v
+		globalconfig.SetServiceName(c.serviceName)
```
Comment (gbbr) [inline ddtrace/tracer/option.go]: I don't know if this is the best place to have this. `defaults` is a function which sets the default for a `*config` parameter. Setting some global state here make assumptions about things which are out of the control of this function, such as: "this function is always called when the tracer starts". We should not make such assumptions. Instead, I'd move this into `tracer.Start` or some better place.

Any thoughts?

File: contrib/Shopify/sarama/option.go [hunk 23]
```diff
@@ -7,15 +7,22 @@ package sarama
...
+	producerServiceName string
+	analyticsRate       float64
 }
 
 func defaults(cfg *config) {
-	cfg.serviceName = "kafka"
+	cfg.producerServiceName = "kafka"
+	cfg.consumerServiceName = "kafka"
+	if svc := globalconfig.ServiceName(); svc != "" {
+		cfg.consumerServiceName = svc
```
Comment (gbbr) [inline contrib/Shopify/sarama/option.go]: Should we drop a comment why we're changing one but not the other?

File: ddtrace/tracer/option.go (hunk 18 above)
Comment (gbbr) [reply ddtrace/tracer/option.go]: FWIW A warning sounds good too!

File: contrib/Shopify/sarama/option.go (hunk 23 above)
Comment (knusbaum) [reply contrib/Shopify/sarama/option.go]: I'm not so sure. I feel that if we want to document this, it might be better to do it in one place for all the integrations.

 I think the people who would find that info useful won't be (or hopefully won't have to be) digging through our code to find it. 

Maybe a line in the commit message about it would be appropriate for any person in the future wondering what the change was about.

File: ddtrace/tracer/option.go (hunk 22 above)
Comment (knusbaum) [reply ddtrace/tracer/option.go]: 👍 I agree.

I had it in `tracer.Start` initially, but it looked weird there to me, away from all the other configuration. It looked to me like the only exported way to start the tracer was `tracer.Start`, and `defaults` is only called from `newUnstartedTracer`. While we have the ability to support multiple tracers, currently it's only exposed as a singleton.

I think you're right, though. This makes `defaults` unsafe to call except when starting a new tracer. I don't know why we'd call it any other time, but it's unexpected that defaults would mutate global state. I think `tracer.Start` is a good place to do it, because that's another place we're already mutating global state (`internal.SetGlobalTracer(newTracer(opts...))`)

File: ddtrace/tracer/option.go (hunk 18 above)
Comment (knusbaum) [reply ddtrace/tracer/option.go]: If I move the `globalconfig.SetServiceName(...)` out of here and into `tracer.Start`, I won't be able to detect this situation to either unset it *or* warn about it.

The best we can do is add a warning any time `WithServiceName` is called about it being deprecated.

File: ddtrace/tracer/option.go (hunk 22 above)
Comment (knusbaum) [reply ddtrace/tracer/option.go]: Actually, we can't move this or it will set the global tracer name when `WithServiceName` is used.
I will move it into `newUnstartedTracer` which isn't perfect, but is slightly better. We can improve this further when we go to v2 and remove the old API.

File: contrib/confluentinc/confluent-kafka-go/kafka/option.go [hunk 24]
```diff
@@ -42,7 +49,8 @@ func WithContext(ctx context.Context) Option {
 // WithServiceName sets the config service name to serviceName.
 func WithServiceName(serviceName string) Option {
 	return func(cfg *config) {
-		cfg.serviceName = serviceName
+		cfg.consumerServiceName = serviceName
+		cfg.producerServiceName = serviceName
```
Comment (jdgumz) [inline contrib/confluentinc/confluent-kafka-go/kafka/option.go]: I get that this preserves the existing behavior (when there was just `cfg.serviceName`). But if someone wants to set a specific service for their producer spans, how would they do so? Do we offer that functionality?

File: ddtrace/tracer/option.go [hunk 25]
```diff
@@ -186,16 +186,27 @@ func WithPropagator(p Propagator) StartOption {
...
+// to WithService, please note that WithService will determine the service name of
+// server and framework integrations. This provides a better experience of the
+// product.
 func WithServiceName(name string) StartOption {
-	return WithService(name)
+	return func(c *config) {
+		c.serviceName = name
+		if globalconfig.ServiceName() != "" {
+			log.Warn("ddtrace/tracer: deprecated config WithServiceName should not be used " +
+				"with `WithService` or `DD_SERVICE` integration service name will not be set.")
```
Comment (jdgumz) [inline ddtrace/tracer/option.go]: nit: semicolon before `integration service name will not be set`?

File: ddtrace/tracer/option.go [hunk 26]
```diff
@@ -186,16 +186,27 @@ func WithPropagator(p Propagator) StartOption {
 	}
 }
 
-// WithServiceName sets the default service name to be used with the tracer. It is
-// deprecated in favour of WithService and will be removed in the next major version.
+// WithServiceName is deprecated. Please use WithService.
+// If you are using an older version and you are upgrading from WithServiceName
+// to WithService, please note that WithService will determine the service name of
+// server and framework integrations. This provides a better experience of the
+// product.
```
Comment (jdgumz) [inline ddtrace/tracer/option.go]: nit: we can probably leave out the last sentence.

File: ddtrace/tracer/option_test.go [hunk 27]
```diff
@@ -114,32 +114,20 @@ func TestTracerOptionsDefaults(t *testing.T) {
...
 			WithSampler(NewRateSampler(0.5)),
-			WithServiceName("api-intake"),
 			WithAgentAddr("ddagent.consul.local:58126"),
 			WithGlobalTag("k", "v"),
 			WithDebugMode(true),
 			WithEnv("testEnv"),
 		)
 		c := tracer.config
 		assert.Equal(float64(0.5), c.sampler.(RateSampler).Rate())
-		assert.Equal("api-intake", c.serviceName)
```
Comment (jdgumz) [inline ddtrace/tracer/option_test.go]: Why the changes to this test case?

File: contrib/confluentinc/confluent-kafka-go/kafka/option.go (hunk 24 above)
Comment (knusbaum) [reply contrib/confluentinc/confluent-kafka-go/kafka/option.go]: No, we don't support using different service names for clients/producers for the same client. It's assumed you will use one client for producer and another for consumer. They can be configured separately and given different service names.

If that's something we want we can open a ticket for it.

File: ddtrace/tracer/option.go (hunk 26 above)
Comment (knusbaum) [reply ddtrace/tracer/option.go]: 👍

File: ddtrace/tracer/option.go (hunk 25 above)
Comment (knusbaum) [reply ddtrace/tracer/option.go]: 👍

File: ddtrace/tracer/option_test.go (hunk 27 above)
Comment (knusbaum) [reply ddtrace/tracer/option_test.go]: This specific test case was moved into the below `TestServiceName/with-service-name` case.
It just duplicates test coverage unnecessarily.

Comment (jdgumz) [reply ddtrace/tracer/option_test.go]: Ah gotcha, okay!

File: ddtrace/tracer/option_test.go [hunk 28]
```diff
@@ -158,3 +146,47 @@ func TestTracerOptionsDefaults(t *testing.T) {
 		assert.Equal(nil, dVal)
 	})
 }
+
+func TestServiceName(t *testing.T) {
+	t.Run("with-service-name", func(t *testing.T) {
```
Comment (gbbr) [inline ddtrace/tracer/option_test.go]:
```suggestion
	t.Run("WithServiceName", func(t *testing.T) {
```

File: ddtrace/tracer/option_test.go [hunk 29]
```diff
@@ -158,3 +146,47 @@ func TestTracerOptionsDefaults(t *testing.T) {
...
+		assert := assert.New(t)
+		tracer := newTracer(
+			WithServiceName("api-intake"),
+		)
+
+		assert.Equal("api-intake", tracer.config.serviceName)
+		assert.Equal("", globalconfig.ServiceName())
+	})
+
+	t.Run("with-service", func(t *testing.T) {
```
Comment (gbbr) [inline ddtrace/tracer/option_test.go]:
```suggestion
	t.Run("WithService", func(t *testing.T) {
```

File: ddtrace/tracer/option_test.go [hunk 30]
```diff
@@ -158,3 +146,47 @@ func TestTracerOptionsDefaults(t *testing.T) {
...
+		assert := assert.New(t)
+		tracer := newTracer(
+			WithService("api-intake"),
+		)
+		assert.Equal("api-intake", tracer.config.serviceName)
+		assert.Equal("api-intake", globalconfig.ServiceName())
+	})
+
+	t.Run("env-service", func(t *testing.T) {
```
Comment (gbbr) [inline ddtrace/tracer/option_test.go]:
```suggestion
	t.Run("env", func(t *testing.T) {
```

File: ddtrace/tracer/option_test.go [hunk 31]
```diff
@@ -158,3 +146,47 @@ func TestTracerOptionsDefaults(t *testing.T) {
...
+		os.Setenv("DD_SERVICE", "api-intake")
+		defer os.Unsetenv("DD_SERVICE")
+		assert := assert.New(t)
+		tracer := newTracer()
+
+		assert.Equal("api-intake", tracer.config.serviceName)
+		assert.Equal("api-intake", globalconfig.ServiceName())
+	})
+
+	t.Run("env-with-option", func(t *testing.T) {
```
Comment (gbbr) [inline ddtrace/tracer/option_test.go]:
```suggestion
	t.Run("combo", func(t *testing.T) {
```

File: ddtrace/tracer/tracer.go [hunk 32]
```diff
@@ -133,6 +134,7 @@ const payloadQueueSize = 1000
 func newUnstartedTracer(opts ...StartOption) *tracer {
 	c := new(config)
 	defaults(c)
+	globalconfig.SetServiceName(c.serviceName)
```
Comment (gbbr) [inline ddtrace/tracer/tracer.go]: With all my shame, I take back my previous comment, this looks quite off place here and seemed much better in `defaults`.
